qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -148,17 +167,220 @@ def _canonicalize(self, dag) -> Tuple[DAGCircuit, bool]:
+
+        # For each qubit, keep the nodes at the start or at the back of the circuit involving this qubit
+        # (also allowing None and empty values).
+        q_to_f = {}
+        q_to_b = {}
+
+        # keep the set of nodes that have been moved either to front_block or to back_block
+        processed_nodes = set()
+
+        # optimization to only consider qubits involved in nodes at the front or at the back of the
+        # circuit
+        unchecked_qubits = set()
+
+        # optimization to keep pairs of nodes for which the inverse check was performed and the nodes
+        # were found to be not inverse to each other
+        checked_node_pairs = set()
+
+        # compute in- and out- degree for every node
+        # also update information for nodes at the start and at the end of the circuit
+        for node in dag.op_nodes():
+            preds = dag.op_predecessors(node)
+            in_degree[node] = len(preds)
+            if len(preds) == 0:
+                for q in node.qargs:
+                    q_to_f[q] = node
+                    unchecked_qubits.add(q)
+            succs = dag.op_successors(node)
+            out_degree[node] = len(succs)
+            if len(succs) == 0:
+                for q in node.qargs:
+                    q_to_b[q] = node
+                    unchecked_qubits.add(q)
+
+        # iterate while there is a possibility to find more inverse pairs
+        while len(unchecked_qubits) > 0:
+            to_check = unchecked_qubits.copy()
+            unchecked_qubits.clear()
+
+            # For each qubit q, check whether the gate at the front of the circuit that involves q
+            # and the gate at the end of the circuit that involves q are inverse
+            for q in to_check:
+
+                if (fn := q_to_f.get(q, None)) is None:
+                    continue
+                if (bn := q_to_b.get(q, None)) is None:
+                    continue
+
+                # fn or bn could be already collected when considering other qubits
+                if fn in processed_nodes or bn in processed_nodes:
+                    continue
+
+                # it is possible that the same node is both at the front and at the back,
+                # it should not be collected
+                if fn == bn:
+                    continue
+
+                # have been checked before
+                if (fn, bn) in checked_node_pairs:
+                    continue
+
+                # fast check based on the arguments
+                if fn.qargs != bn.qargs or fn.cargs != bn.cargs:
+                    continue
+
+                # in the future we want to include a more precise check whether a pair
+                # of nodes are inverse
+                if fn.op == bn.op.inverse():
+                    # update q_to_f, q_to_b maps
+                    for q in fn.qargs:
+                        q_to_f[q] = None
+                    for q in bn.qargs:
+                        q_to_b[q] = None
```

Comment:
I think that if we used `del q_to_f[q]` and so on, we might not need `unchecked_qubits` at all - we could use `to_check = q_to_f.keys() & q_to_b.keys()`.

Reply:
Agreed. I was thinking of doing this, but was worried that deleting keys would be slower than recomputing the active qubits at each iteration. Though I am no longer sure this is really the case.